File: core/parser/singbox.py
"""
Парсер родногo формата конфигураций sing-box (JSON).
"""

import json
from typing import Any, Dict, List

from core.config.models import ProfileConfig, ProtocolType
from core.parser.base import BaseParser, ParseResult
from utils.logger import log_call


class SingBoxParser(BaseParser):
    """Парсер для native sing-box JSON конфигураций."""

    format_name: str = "sing-box"

    TYPE_MAP = {
        "shadowsocks": ProtocolType.SHADOWSOCKS,
        "vmess": ProtocolType.VMESS,
        "vless": ProtocolType.VLESS,
        "trojan": ProtocolType.TROJAN,
        "hysteria2": ProtocolType.HYSTERIA2,
        "wireguard": ProtocolType.WIREGUARD,
        "tuic": ProtocolType.TUIC,
        "ssh": ProtocolType.SSH,
        "direct": ProtocolType.DIRECT,
        "block": ProtocolType.BLOCK,
        "dns": ProtocolType.DNS,
    }

    @log_call
    def parse(self, content: str) -> ParseResult:
        """
        Разбирает JSON-конфигурацию sing-box.

        Args:
            content: Текст JSON sing-box.

        Returns:
            ParseResult: Результат парсинга.
        """
        result = ParseResult(format_name=self.format_name)

        try:
            data = json.loads(content)
        except json.JSONDecodeError as exc:
            result.errors.append(f"Ошибка синтаксиса sing-box JSON: {exc}")
            return result

        if not isinstance(data, dict):
            result.errors.append("Корневой элемент sing-box JSON должен быть словарем.")
            return result

        result.raw_config = data
        outbounds = data.get("outbounds", [])

        if not isinstance(outbounds, list):
            result.errors.append("Секция 'outbounds' должна быть списком.")
            return result

        for item in outbounds:
            if not isinstance(item, dict):
                continue

            out_type = item.get("type", "").lower()
            tag = item.get("tag", "singbox-outbound")

            # Пропускаем служебные outbounds
            if out_type in ("dns", "direct", "block"):
                continue

            server = item.get("server", "127.0.0.1")
            server_port = int(item.get("server_port", 443))

            protocol = self.TYPE_MAP.get(out_type, ProtocolType.VLESS)

            # Сохраняем остальные поля в settings
            settings = {k: v for k, v in item.items() if k not in ("type", "tag", "server", "server_port")}

            profile = ProfileConfig(
                name=tag,
                protocol=protocol,
                server=server,
                server_port=server_port,
                settings=settings,
                tag=tag,
            )
            result.profiles.append(profile)

        return result

File: core_manager/src/parser/singbox.rs
//! Парсер native sing-box JSON конфигураций на Rust.

use crate::models::{ProfileConfig, ProtocolType};
use crate::parser::{ConfigParser, ParseResult};

pub struct SingBoxParser;

impl ConfigParser for SingBoxParser {
    fn format_name(&self) -> &'static str {
        "sing-box"
    }

    fn parse(&self, content: &str) -> ParseResult {
        let mut result = ParseResult::new(self.format_name());

        let json_val: serde_json::Value = match serde_json::from_str(content) {
            Ok(v) => v,
            Err(err) => {
                result.errors.push(format!("Ошибка синтаксиса sing-box JSON: {}", err));
                return result;
            }
        };

        result.raw_config = Some(json_val.clone());

        if let Some(outbounds) = json_val.get("outbounds").and_then(|o| o.as_array()) {
            for item in outbounds {
                let out_type = item.get("type").and_then(|t| t.as_str()).unwrap_or("vless").to_lowercase();
                let tag = item.get("tag").and_then(|t| t.as_str()).unwrap_or("singbox-proxy");

                if ["direct", "block", "dns"].contains(&out_type.as_str()) {
                    continue;
                }

                let server = item.get("server").and_then(|s| s.as_str()).unwrap_or("127.0.0.1");
                let server_port = item.get("server_port").and_then(|p| p.as_u64()).unwrap_or(443) as u16;

                let protocol = match out_type.as_str() {
                    "shadowsocks" => ProtocolType::Shadowsocks,
                    "vmess" => ProtocolType::Vmess,
                    "vless" => ProtocolType::Vless,
                    "trojan" => ProtocolType::Trojan,
                    "hysteria2" => ProtocolType::Hysteria2,
                    "wireguard" => ProtocolType::Wireguard,
                    "tuic" => ProtocolType::Tuic,
                    "ssh" => ProtocolType::Ssh,
                    "direct" => ProtocolType::Direct,
                    "block" => ProtocolType::Block,
                    _ => ProtocolType::Vless,
                };

                let mut p = ProfileConfig::new(tag, protocol, server, server_port);
                p.tag = tag.to_string();
                p.settings = item.clone();
                result.profiles.push(p);
            }
        }

        result
    }
}
